### ISSUE TITLE
Mod download progress bar won't move, some mods aren't updating, subscribed mods aren't showing in the mods menu

### ISSUE THREAD

**Issue by Krzeszny** [Regalis11/Barotrauma#11325]
### Disclaimers

- [X] I have searched the issue tracker to check if the issue has already been reported.
- [X] My issue happened while using mods.

### What happened?
I know there must be a report for it somewhere here, but I couldn't find any similar issues using keywords.

![Barotrauma_8nffuqqJVn](https://user-images.githubusercontent.com/10529134/221026926-f7da12a5-1864-426e-8fd2-0ebaceef126c.png)
The issue is that some mods won't update and can't be reinstalled because after a failed install, they won't show in the Mods menu.
I keep getting this issue almost every time a few mods update, sometimes when one mod gets an update.

Today, I had to update Neurotrauma. I couldn't. Then I subscribed to more mods while trying to join a server but another mod wouldn't update just like Neurotrauma. When I  clicking Cancel and check the mods list, even after restarting the game, the mods that have failed to update aren't visible on the list.

What is causing this issue? Is it the modding refactor? Is it Steam? How to prevent or fix it? It happens randomly albeit regularly, even if I keep removing the mods folder once in a week. It's really annoying.

### Bug prevalence

Happens regularly

### Version

0.21.6.0

### Which operating system did you encounter this bug on?

Windows

**Comment by Krzeszny** [comment]
![Barotrauma_Mzu0Woigx5](https://user-images.githubusercontent.com/10529134/221373094-889fdfd5-2481-4a8f-b964-b3f297868baa.png)
They just won't update

**Comment by Krzeszny** [comment]
Today, with less than 10 mods to update, I had to click cancel and re-update 4 or 5 times.
It goes like this:
One progress bar fills and update stops. Click cancel. Click update. Another progress bar fills and update stops. Cancel. Update. etc.

**Comment by Krzeszny** [comment]
Not present in v1.0.7.0, so I'm closing it as fixed.

**Comment by Krzeszny** [comment]
Spoke too soon. It doesn't happen on updating but I've spotted it 2 times while subscribing to new mods. Example:
(it won't finish)
![Barotrauma_OdshCOjkPS](https://user-images.githubusercontent.com/10529134/225142064-4a41281c-631d-4b29-a15f-9b1ac699ff3b.png)

**Comment by Krzeszny** [comment]
![Barotrauma_e6ijSDnR8q](https://user-images.githubusercontent.com/10529134/225756343-9082a512-efb4-4aff-bcbf-afe5aa67c285.png)
Is it just me?

**Comment by Krzeszny** [comment]
It happens practically every day. I've found a possible workaround: unsubscribe and restart.

As a semi-fix, can you make it so when the progress bar freezes for some reason, there's a button that says "retry" or "reinstall unfinished downloads"?

**Comment by Krzeszny** [comment]
Same thing: updating progress won't resume.
![Barotrauma_zFVT4GP7KI](https://github.com/Regalis11/Barotrauma/assets/10529134/3d7f515d-4e34-461b-89a5-f8bdcf9d4c9d)

**Comment by Krzeszny** [comment]
@Regalis11 I've done some lengthy testing. I've found and compared 12 mismatched mods between the AppData and the Steam workshop paths.

For that purpose, I've made copies of each folder, tagged them properly and used WinMerge. For 10 of those mods, the only differing file between Workshop and AppData was filelist.xml (specifically the mod's name and the attributes), while for the other two mods, empty folder(s) were missing for the 2 mods in AppData. Then I reinstalled a few mods using the Mods menu by right-clicking and pressing Reinstall. I did it with the 2 mods which were missing folders and a few other mods just for comparison, as the results seemed the same for each one: even after reinstalling, filelist.xml was still different between the 2 directories. 

However, in the two cases where other folders were missing besides the filelist.xml differences, the missing folders were added to AppData upon reinstalling. After reinstallation, I still had 12 mods with mismatched filelist.xml files.

One example of mismatched mods was EK: Armory:
Before reinstalling, besides the missing 2 empty folders in AppData (named Health and Sound), these were the differing lines in filelist.xml:
AppData pre-reinstall: 
`<contentpackage name="EK | Armory" steamworkshopid="1769403760" corepackage="false" modversion="1.0.6" gameversion="0.19.11.0" installtime="1665076052" altnames="EK Armory" expectedhash="991ECA75172CB240D26DECFBCB705127">'
AppData reinstalled: 
`<contentpackage name="EK Armory" steamworkshopid="1769403760" corepackage="false" modversion="1.0.6" gameversion="0.19.11.0" installtime="1665076052" altnames="EK Armory" expectedhash="991ECA75172CB240D26DECFBCB705127">`
Workshop pre-reinstall: 
'<contentpackage name="EK Armory" modversion="1.0.6" corepackage="False" steamworkshopid="1769403760" gameversion="0.19.11.0" expectedhash="991ECA75172CB240D26DECFBCB705127">`expectedhash="991ECA75172CB240D26DECFBCB705127">`
Workshop reinstalled: 
no differences

As you can see, the Workshop folder file stayed the same, but the AppData file changed slightly (by removing ` |` from the name, but still remained mismatched.


A second example where another empty folder was missing from the AppData directory is Combinable Ammo Revived:
AppData pre-reinstall:
`<contentpackage name="Combinable Ammo Revived (NOW WORKS WITH ALL AMMO)" steamworkshopid="2799051561" corepackage="false" modversion="1.0.6" gameversion="0.20.16.1" installtime="1680387627" altnames="CombinableCoilgunAmmoRevived" expectedhash="32F31BEB0252EF673F510BFA65B74995">`
AppData reinstalled:
`<contentpackage name="CombinableCoilgunAmmoRevived" steamworkshopid="2799051561" corepackage="false" modversion="1.0.6" gameversion="0.20.16.1" installtime="1680387627" altnames="CombinableCoilgunAmmoRevived" expectedhash="32F31BEB0252EF673F510BFA65B74995">`
Workshop pre-reinstall:
`<contentpackage name="CombinableCoilgunAmmoRevived" modversion="1.0.6" corepackage="False" steamworkshopid="2799051561" gameversion="0.20.16.1" expectedhash="32F31BEB0252EF673F510BFA65B74995">
Workshop reinstalled:
no changes

Again, the AppData file changed but remained mismatched, while the workshop file remained the same. 
In both cases presented the mod's name had changed.

The same happened with T.I.T.A.N. Warsuit, where the AppData file was outdated, then reinstalling updated the mod's name but the file remained mismatched because the attributes were placed differently than in the workshop file. We can therefore assume that the same happens with every mod: reinstalling doesn't fix the issue.

Then I tried unsubscribing from Barotrauma's mod menu. After unsubbing, the mods weren't there in either directory, as Barotrauma removed both folders. So I subscribed manually and... the folders reappeared and filelist.xml was still mismatched in the same way as after reinstalling. How is that possible? Where does the game get the mismatched filelist.xml from, even if the folder disappears from AppData? Somehow, the game changes the attributes in the AppData version. Even the altname attrib differs.

For comparison, I unsubbed a mismatched mod with Barotrauma closed. As expected, the folder in AppData remained there. Then I relaunched the game and opened settings. The folder disappeared. I closed the game, subbed the mod and relaunched the game, to find filelist.xml mismatched again: the attributes were still in wrong order, `altname` differed and the AppData version had an additional `installtime` attribute. I don't even know if it matters or not, but technically the files aren't identical.

The last thing I did with the mismatched mods was removing both mod folders manually from AppData and Workshop folders and launching Barotrauma. The result: filelist.xml was STILL mismatched, with different/missing/additional attributes. An example after removing the folders manually and launching Baro, then opening Mods:
AppData:
`<contentpackage name="T.I.T.A.N. power suit V2" steamworkshopid="2346376423" corepackage="false" modversion="1.0.11" gameversion="1.0.8.0" installtime="1679959957" altnames="T.I.T.A.N. power suit V2" expectedhash="70E69E948BF09B9919688410981430BE">`
Workshop:
`<contentpackage name="T.I.T.A.N. power suit V2" modversion="1.0.11" corepackage="False" steamworkshopid="2346376423" gameversion="1.0.8.0" altnames="T.I.T.A.N. power suit - New update,New me." expectedhash="70E69E948BF09B9919688410981430BE">`

The attributes differ significantly.
![image](https://github.com/Regalis11/Barotrauma/assets/10529134/da851063-b078-4374-8322-53c4d3f729ff)


Out of curiosity, I compared a random mod between AppData and Workshop, which turned out to be Instrumental Music mod. The version from AppData was converted to the %ModDir% format, while the Workshop folder version had the outdated Mods/ format. This is probably intended.

I compared a second random mod, which was Visual Variety Pack. The version from AppData had an `installtime` attribute, while the Workshop version didn't have it. Both had the %ModDir% format.

Another random mod, SirenMod also was converted to %ModDir% in AppData. It must be intentional.

A fourth random mod was Flintlock Rifle, which I subscribed to today for the first time. The AppData version also had an `installtime` attribute, unlike the Workshop version. I think this difference may be intentional, after all.

To sum up, Barotrauma messes up attributes somehow, changing their values and their order, even adding new attributes in AppData. I haven't got a clue if it's intended or not, but at least after resubscribing, a mismatched mod's name changes to the updated name. I hope you can get some useful info from this comment.

**Comment by Krzeszny** [comment]
Also, here's a list of what mods I've compared and what the differences were between AppData and Workshop (without reinstalling). 
Missing is for AppData. installtime is added for AppData.

EK Armory: 2 missing empty folders, different name, different altnames, installtime, attribute order
Combiname Ammo: 1 missing empty folder, different name and altnames, installtime, attribute order
EK Dockyard: different name and altnames, installtime, attribute order
Exosuit Retexture: installtime, attribute order
Genes+ Revived and Expanded: different name and altnames, installtime, attribute order
EK Gunnery: different name and altnames, installtime, attribute order
Husk Speak & Spell: installtime, attribute order
Neurotrauma Tutorial map Mercy Hospital: different name and altnames, installtime, attribute order
No Role Balance v2: installtime, attribute order
NT Cybernetics: different altnames (but not name), installtime="2022Y 04M 03D 03HR 26MIN 53SEC UTC" in the workshop but just numbers in AppData, attribute order
T.I.T.A.N. power suit V2: different name and altnames, installtime, attribute order
EK Utilities: different name and altnames, installtime, attribute order

**Comment by Regalis11** [comment]
Thank you for taking the time to test this and the very thorough report!

What's happening here is that Steam downloads the mod files as-is, and when the game installs them to the AppData folder, it resaves the xml, filling in details such as the install time and any other missing attributes. This can also cause the order of the attributes to change. However, this should not cause issues with joining the server: as long as the MD5 hashes and contents of the mod are the same, the differences in the attributes don't matter.

Since it seems like you're using some kind of source code mods, and since this doesn't seem to be such a common issue in the vanilla game (I don't recall us getting any other reports about it), I have to wonder if this could have something to do with the mods you're using?

I also feel like I might be misunderstanding the issue somehow, so to be clear:
- Are you downloading the mods from the server, or from the Workshop? I.e. when you get the prompt about missing mods, do you select that you'd like to download them from the server, or that you'd like to subscribe to them in the workshop? I had always assumed you're trying to download them, but your recent tests seem to suggest you're trying to get them from the Workshop (which only works if the server is using up-to-date versions of the mods from the Workshop!)
- Are you actually sure the mods don't download? To me it looks that all the downloads have finished, but the prompt hasn't closed for some reason.
- It's possible to get some additional debug info about the download progress by entering the command "verboselogging" in the console. Using this _might_ give some extra clues on what's going wrong here.

But in any case, I'm sorry we still haven't been able to get to the bottom of this, I can imagine how frustrating it must be to keep constantly running into the same problem. :/

**Comment by Regalis11** [comment]
...I think I finally managed to reproduce this now! The issue was that I thought this was about downloading the mods from the server. This is why accurate reproduction steps are important. 🙃 Thank you for the extra information you provided, I think I would've kept trying to debug downloading mods from the server otherwise.

**Comment by Krzeszny** [comment]
> ...I think I finally managed to reproduce this now! The issue was that I thought this was about downloading the mods from the server. This is why accurate reproduction steps are important. 🙃 Thank you for the extra information you provided, I think I would've kept trying to debug downloading mods from the server otherwise.

Yup, sorry for not enough details. All of the information I provided was about subscribing to mods while joining a server, not downloading them directly. The biggest problem is when a mod author updates the mod's name or leaves empty folders. Then Barotrauma doesn't update the AppData filelist.xml properly or doesn't remove empty folders and that's probably what causes a mismatch.

However, as you've read, there are mismatches detected even with mods that haven't been updated for months. I suspect it has to do with the name having been changed long ago and the host having an outdated AppData filelist.xml despite my own files being correct. So all in all, I blame outdated AppData filelist.xml.

>Are you actually sure the mods don't download? To me it looks that all the downloads have finished, but the prompt hasn't closed for some reason.
I'm gonna check it using `verboselogging`.

>However, this should not cause issues with joining the server: as long as the MD5 hashes and contents of the mod are the same, the differences in the attributes don't matter.
The MD5 hash isn't calculated when joining a server, right? Otherwise, if the filelist.xml content is different because of the attributes, I'd expect a different hash. But since ALL AppData mods have the installtime attrib, it has to be normal.

**Comment by Regalis11** [comment]
I think I managed to find and fix the issue now. Should be fixed in https://github.com/Regalis11/Barotrauma-development/commit/a062d203d8e36d619eba6dade0ef4bd1f80b09ea. 

> Fixed workshop item downloads sometimes getting stuck when opting to install server mods from the workshop. It seems the InstallWaiter was left running indefinitely if the mod had already been installed, if it was already in the process of being installed, or if it couldn't be installed. Also added text on the download progressbar to indicate the state of the install.

Thank you so much for the help in diagnosing this!

> However, as you've read, there are mismatches detected even with mods that haven't been updated for months. I suspect it has to do with the name having been changed long ago and the host having an outdated AppData filelist.xml despite my own files being correct. So all in all, I blame outdated AppData filelist.xml.

I don't think this is related, the only thing that should make a difference when determining compatibility is the MD5 hash, and those seem to match in all of these cases. 

> What I still don't understand is why the AppData version of T.I.T.A.N. Warsuit has a different altname attribute value, even after reinstalling.

This happens because we used to use the title of the workshop as an "alternative name", but currently the title isn't used for anything but displaying the name of the mod in the Workshop.

**Comment by Regalis11** [comment]
Internal note for QA: this seems to be fairly easily reproducible by joining servers that have mods you haven't subscribed to, and then opting to subscribe to them when you get the prompt about missing mods.

**Comment by Krzeszny** [comment]
> Fixed workshop item downloads sometimes getting stuck when opting to install server mods from the workshop. It seems the InstallWaiter was left running indefinitely if the mod had already been installed, if it was already in the process of being installed, or if it couldn't be installed. Also added text on the download progressbar to indicate the state of the install.

Just to confirm, it used to happen because when a mod is mismatched (a separate issue), the server still asks the client to subscribe to it, even if the client is subscribed, correct?

**Comment by Regalis11** [comment]
> Just to clarify, it used to happen because when a mod is mismatched (a separate issue), the server still asks the client to subscribe to it, even if the client is subscribed, correct?

Basically yeah, or in a situation where the mod is already being installed, or if it can't be installed for whatever reason.

**Comment by Krzeszny** [comment]
Just some feedback. 
When I clicked Update in the main menu today, Steam decided to validave 4.8 GB of workshop files. (Imagine the if you had to do it with an HDD, lol). When validation happens in-game like that, Barotrauma doesn't show any update progress until it finishes and a user can think that it's not doing anything and click Cancel.

**Comment by Krzeszny** [comment]
It's just happened for me (current Baro version) while Steam wasn't downloading anything in the background. The downloading just froze and I had to cancel it after a minute.
![Barotrauma_L92CJdDaIj](https://github.com/Regalis11/Barotrauma/assets/10529134/9e60ec36-bfc5-49a5-a519-64a86ae44dcf)

**Comment by Krzeszny** [comment]
Same as in the last comment. Verboselogging shows nothing. The download wouldn't finish on 3 attempts, but on the 4th try it finished and I joined the server without having to download mods directly. What a weird glitch.
![Barotrauma_FnjUZrVKlV](https://github.com/Regalis11/Barotrauma/assets/10529134/5ab71407-70c6-48a6-8d60-41fc30375146)

**Comment by Regalis11** [comment]
@Krzeszny Yeah, the fix isn't yet included in either the unstable or the stable version. We're hoping to get it in the next update though. Sorry for the inconvenience!
